src/cache/cache-manager.ts

```diff
@@ -0,0 +1,152 @@
+import { CacheKeyManifest } from './key-manifest';
+
+import {
+  CacheEntry,
+  ICache,
+  CacheKey,
+  CACHE_KEY_PREFIX,
+  WrappedCacheEntry
+} from './shared';
+
+const DEFAULT_EXPIRY_ADJUSTMENT_SECONDS = 0;
+
+export class CacheManager {
+  constructor(
+    private cache: ICache,
+    private readonly keyManifest: CacheKeyManifest
+  ) {}
+
+  async get(
+    cacheKey: CacheKey,
+    expiryAdjustmentSeconds = DEFAULT_EXPIRY_ADJUSTMENT_SECONDS
+  ): Promise<Partial<CacheEntry> | undefined> {
+    let wrappedEntry = await this.cache.get<WrappedCacheEntry>(
+      cacheKey.toKey()
+    );
+
+    if (!wrappedEntry) {
+      // Try again by loosely-matching the key against a set of keys we've stored
+      // in the key manifest.
+      const keyManifestEntry = await this.keyManifest.get();
+
+      if (keyManifestEntry) {
+        const matchedKey = this.matchExistingCacheKey(
+          cacheKey,
+          keyManifestEntry.keys
+        );
+        wrappedEntry = await this.cache.get<WrappedCacheEntry>(matchedKey);
+      }
+
+      // If we still don't have an entry..
+      if (!wrappedEntry) {
+        return;
+      }
+    }
+
+    const nowSeconds = Math.floor(Date.now() / 1000);
+
+    // Make sure the key manifest knows about the key.
+    // This helps to migrate keys into the manifest, as the manifest takes care
+    // of duplicates for us.
+    await this.keyManifest.add(cacheKey.toKey());
```

Comment:
I think there are some edge case issues with this migration, eg

If `localStorage` has:

```
foo: 1
bar baz: 2
```

`cache.get('foo')` will exact match and populate the manifest with `['foo']` then return `1`

Then a call to `cache.get('bar')` will not exact match, `bar baz` wont be in the manifest, so would return `undefined`, when it should return `2`

Annoyingly, you either need to move all the keys in at once or not use the manifest for `localStorage` at all (which might be the safer option)

Reply:
> not use the manifest for localStorage at all

I went with this option. Additionally, we could also use it for the in-memory cache.